src/PowerShellEditorServices/Services/PowerShell/Console/UnixConsoleOperations.cs

```diff
@@ -0,0 +1,298 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PowerShell.EditorServices.Utility;
+using UnixConsoleEcho;
+
+namespace Microsoft.PowerShell.EditorServices.Services.PowerShell.Console
+{
+    internal class UnixConsoleOperations : IConsoleOperations
+    {
+        private const int LongWaitForKeySleepTime = 300;
+
+        private const int ShortWaitForKeyTimeout = 5000;
+
+        private const int ShortWaitForKeySpinUntilSleepTime = 30;
+
+        private static readonly ManualResetEventSlim s_waitHandle = new ManualResetEventSlim();
+
+        private static readonly SemaphoreSlim s_readKeyHandle = AsyncUtils.CreateSimpleLockingSemaphore();
+
+        private static readonly SemaphoreSlim s_stdInHandle = AsyncUtils.CreateSimpleLockingSemaphore();
+
+        private Func<CancellationToken, bool> WaitForKeyAvailable;
+
+        private Func<CancellationToken, Task<bool>> WaitForKeyAvailableAsync;
+
+        internal UnixConsoleOperations()
+        {
+            // Switch between long and short wait periods depending on if the
+            // user has recently (last 5 seconds) pressed a key to avoid preventing
+            // the CPU from entering low power mode.
+            WaitForKeyAvailable = LongWaitForKey;
+            WaitForKeyAvailableAsync = LongWaitForKeyAsync;
+        }
+
+        public ConsoleKeyInfo ReadKey(bool intercept, CancellationToken cancellationToken)
+        {
+            s_readKeyHandle.Wait(cancellationToken);
+
+            // On Unix platforms System.Console.ReadKey has an internal lock on stdin.  Because
+            // of this, if a ReadKey call is pending in one thread and in another thread
+            // Console.CursorLeft is called, both threads block until a key is pressed.
+
+            // To work around this we wait for a key to be pressed before actually calling Console.ReadKey.
+            // However, any pressed keys during this time will be echoed to the console. To get around
+            // this we use the UnixConsoleEcho package to disable echo prior to waiting.
+            if (VersionUtils.IsPS6)
+            {
+                InputEcho.Disable();
```

Comment:
This has been fixed in PowerShell 7 because .NET 3.1 fixed it, hence we can remove this code now.